### PR TITLE
allow flexible log parsing and formatting

### PR DESCRIPTION
**Pull request by niamster**
Addresses https://github.com/stern/stern/issues/227

This is an alternative solution to https://github.com/stern/stern/pull/228

Instead of hardcoding template in the code, I suggest adding a bunch of helper functions and a way to load template from disk.

This PR adds:
 * various color functions
 * TS format function `formatTsRFC3339Nano`
 * `tryParseJSON` function that does not return an error
 * `--template-file/-T` to load template from disk

Sample template that I use:
```
{{color .PodColor .PodName}} {{color .ContainerColor .ContainerName}} {{ with $msg := .Message | tryParseJSON }}[{{ colorGreen (formatTsRFC3339Nano $msg.ts) }}] {{ levelColor $msg.level }} ({{ colorCyan $msg.caller }}) {{ $msg.msg }}{{ else }} {{ .Message }} {{ end }}
```

It allows me to have an alias `alias stern="stern -T ~/.stern.tpl"` which support parsing both JSON and non-JSON logs.

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -302,6 +306,7 @@ func (o *options) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVarP(&o.since, "since", "s", o.since, "Return logs newer than a relative duration like 5s, 2m, or 3h.")
 	fs.Int64Var(&o.tail, "tail", o.tail, "The number of lines from the end of the logs to show. Defaults to -1, showing all logs.")
 	fs.StringVar(&o.template, "template", o.template, "Template to use for log lines, leave empty to use --output flag.")
+	fs.StringVarP(&o.templateFile, "template-file", "T", o.templateFile, "Path to template to use for log lines, leave empty to use --output flag.")
```
Comment (superbrothers) [inline cmd/cmd.go:309]: Please modify so that `--template` and `--template-file` options cannot be used at the same time.

File: README.md [hunk 2]
```diff
@@ -95,6 +95,7 @@ Supported Kubernetes resources are `pod`, `replicationcontroller`, `service`, `d
  `--since`, `-s`             | `48h0m0s` | Return logs newer than a relative duration like 5s, 2m, or 3h.
  `--tail`                    | `-1`      | The number of lines from the end of the logs to show. Defaults to -1, showing all logs.
  `--template`                |           | Template to use for log lines, leave empty to use --output flag.
+ `--template-file`, `-T`     |           | Path to template to use for log lines, leave empty to use --output flag.
```
Comment (superbrothers) [inline README.md:98]: Please add examples using this change to https://github.com/stern/stern#examples.

**Review by superbrothers — COMMENTED** [review]
Please add enough tests as needed for changes to `generateTemplat` function.

File: cmd/cmd.go [hunk 3]
```diff
@@ -348,6 +360,13 @@ func (o *options) generateTemplate() (*template.Template, error) {
 			}
 			return string(b), nil
 		},
+		"tryParseJSON": func(text string) map[string]interface{} {
```
Comment (tksm) [inline cmd/cmd.go:363]: Could you add the new functions to the "templates" section in README.md?
https://github.com/stern/stern#templates

File: cmd/cmd.go [hunk 4]
```diff
@@ -348,6 +360,13 @@ func (o *options) generateTemplate() (*template.Template, error) {
...
+			obj := make(map[string]interface{})
+			if err := json.Unmarshal([]byte(text), &obj); err != nil {
+				return nil
+			}
+			return obj
+		},
```
Comment (tksm) [inline cmd/cmd.go:369]:
```suggestion
		"tryParseJSON": func(text string) map[string]interface{} {
			decoder := json.NewDecoder(strings.NewReader(text))
			decoder.UseNumber()
			obj := make(map[string]interface{})
			if err := decoder.Decode(&obj); err != nil {
				return nil
			}
			return obj
		},
```

Though parseJSON has the same issue, it seems that tryParseJSON does not parse large numbers such as timestamps as expected.

Let's say we have logs below. The timestamp `1257894000` is `2009-11-10 23:00:00` in UTC.

```
{"level":"info","ts":1257894000,"caller":"sandbox2210433069/prog.go:10","msg":"hello","key1":"val1"}
{"level":"info","ts":1257894000,"caller":"sandbox2210433069/prog.go:13","msg":"world","key1":"val1"}
```

When we use `tryParseJSON` and `formatTsRFC3339Nano`, we will get `0001-01-01T00:00:00Z`, which is incorrect.

```
zap-log c [0001-01-01T00:00:00Z] info (sandbox2210433069/prog.go:10) hello
zap-log c [0001-01-01T00:00:00Z] info (sandbox2210433069/prog.go:13) world
```

Please refer to https://stackoverflow.com/questions/22343083/json-unmarshaling-with-long-numbers-gives-floating-point-number for details.

File: cmd/cmd.go [hunk 5]
```diff
@@ -365,9 +384,41 @@ func (o *options) generateTemplate() (*template.Template, error) {
 			}
 			return strings.TrimSuffix(string(b), "\n"), nil
 		},
+		"formatTsRFC3339Nano": func(ts any) string {
```
Comment (tksm) [inline cmd/cmd.go:387]: I think the `Ts` part does not give additional information for users, so what do you think about simply naming it `rfc3339Nano` or `toRFC3339Nano`?

```
{{ colorGreen (rfc3339Nano $msg.ts) }}
```

File: cmd/cmd.go (hunk 4 above)
Comment (niamster) [reply cmd/cmd.go:369]: Thank you for the suggestion, I didn't realise that `UseNumber` is not a default option!

File: cmd/cmd.go (hunk 5 above)
Comment (niamster) [reply cmd/cmd.go:387]: Sounds good to me, I don't have a strong opinion here 👍

File: cmd/cmd.go (hunk 3 above)
Comment (niamster) [reply cmd/cmd.go:363]: Definitely, thank you for pointing it out 🙇

File: cmd/cmd.go (hunk 1 above)
Comment (niamster) [reply cmd/cmd.go:309]: Good point, I wanted to add that one overrides the other but that slipped from my mind.

**Comment by niamster** [comment]
@tksm @superbrothers Thank you so much for the review!
I tried to address your comments, please let me know if I missed something.

NOTE: I added `toUTC` function to simplify testing. It still might be useful in some case I think.

File: cmd/cmd.go (hunk 5 above)
Comment (tksm) [reply cmd/cmd.go]: Thanks for changing! Could you add the functions such as `toRFC3339Nano` to the table in the templates section?


image

File: README.md [hunk 6]
```diff
@@ -247,6 +248,18 @@ Output using a custom template with `parseJSON`:
 stern --template='{{.PodName}}/{{.ContainerName}} {{with $d := .Message | parseJSON}}[{{$d.level}}] {{$d.message}}{{end}}{{"\n"}}' backend
 ```
 
+Output using a custom template that tries to parse JSON or fallbacks to raw format:
+
+```
+stern --template='{{.PodName}}/{{.ContainerName}} {{ with $msg := .Message | tryParseJSON }}[{{ colorGreen (toRFC3339Nano (toUTC $msg.ts)) }}] {{ levelColor $msg.level }} ({{ colorCyan $msg.caller }}) {{ $msg.msg }}{{ else }} {{ .Message }} {{ end }}{{"\n"}}' backend
```
Comment (tksm) [inline README.md:254]:
```suggestion
stern --template='{{.PodName}}/{{.ContainerName}} {{ with $msg := .Message | tryParseJSON }}[{{ colorGreen (toRFC3339Nano $msg.ts) }}] {{ levelColor $msg.level }} ({{ colorCyan $msg.caller }}) {{ $msg.msg }}{{ else }} {{ .Message }} {{ end }}{{"\n"}}' backend
```

Though `toUTC` is helpful for testing, using local time for readability might be better here.

**Comment by tksm** [comment]
Thank you for the changes! I added minor comments below, but it looks good to me other than these. Could you rebase it to allow CI to check?

- https://github.com/stern/stern/pull/239#discussion_r1125155115
- https://github.com/stern/stern/pull/239#discussion_r1125160096

File: cmd/cmd.go (hunk 5 above)
Comment (niamster) [reply cmd/cmd.go]: 👍 adding all of them them

**Comment by niamster** [comment]
> Thank you for the changes! I added minor comments below, but it looks good to me other than these. Could you rebase it to allow CI to check?
> 
> * [allow flexible log parsing and formatting #239 (comment)](https://github.com/stern/stern/pull/239#discussion_r1125155115)
> * [allow flexible log parsing and formatting #239 (comment)](https://github.com/stern/stern/pull/239#discussion_r1125160096)

Thank you so for the suggestions! I merged `master` into my branch. D you want me to rebase instead (TBH I'm not a huge fan of rebase since it messes up with GH quite often, so we loose essential traces)?

**Comment by tksm** [comment]
> Thank you so for the suggestions! I merged master into my branch. D you want me to rebase instead (TBH I'm not a huge fan of rebase since it messes up with GH quite often, so we loose essential traces)?

Thank you. Either rebase or merge is fine with me.

**Review by tksm — APPROVED** [review]
LGTM 👍

**Pull request merged**